controllers/volumereplicationgroup_controller.go

```diff
@@ -201,35 +207,78 @@ func (v *VolumeReplicationGroupReconciler) Reconcile(ctx context.Context, req ct
 		return ctrl.Result{}, fmt.Errorf("failed to reconcile VolumeReplicationGroup %w", err)
 	}
 
-	log.Info("Processing VolumeReplicationGroup ", volRepGroup.Spec, " in ns: ", req.NamespacedName)
+	v.ctx = ctx
+
+	if err := v.updatePVCList(); err != nil {
+		log.Error(err, "Handling of Persistent Volume Claims of application failed")
 
-	// examine the DeletionTimestamp to determine whether the object is under deletion
-	if !volRepGroup.GetDeletionTimestamp().IsZero() {
-		if err := v.runAndRemoveFinalizer(ctx, volRepGroup, vrgFinalizerName); err != nil {
-			log.Errorf("failed to remove the finalizer from vrg: %s/%s", volRepGroup.Name, volRepGroup.Namespace)
+		return ctrl.Result{}, err
+	}
 
-			return ctrl.Result{Requeue: true}, nil
-		}
+	log.Info("Processing VolumeReplicationGroup ", v.instance.Spec, " in ns: ", req.NamespacedName)
 
-		// stop reconciliation as the item is being deleted
+	switch {
+	case !v.instance.GetDeletionTimestamp().IsZero():
+		return v.finalizeVRG()
+	case v.instance.Spec.ReplicationState == volrep.ReplicationPrimary:
+		return v.processAsPrimary()
+	case v.instance.Spec.ReplicationState == volrep.ReplicationSecondary:
+		return v.processAsSecondary()
+	}
+
+	// No requeue as there is no reconcile till user changes desired spec to a valid value
+	return ctrl.Result{Requeue: false},
```

Comment:
(nit) maybe drop the {Requeue: false} given that you are returning an error

Reply:
ack! (also, jFYI this is only compile tested as of now)